File: src/main/browser-history.ts
import { existsSync, mkdirSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import type { IpcMain } from 'electron'
import { writeFileAtomic } from './atomic-write'

/**
 * Where this browser has been — one list per profile, on disk, clickable.
 *
 * ## Why there was none
 *
 * There was no history store, no history UI and no `browser-history:*` channel
 * in any build before this one. `history` in this subsystem only ever meant
 * `webContents.navigationHistory` — the Back and Forward buttons, which are
 * per-tab and die with the tab. Asad read Chrome's ⋮ menu out loud with ours
 * open beside it:
 *
 *   > *"I need most of them, and passwords history also."*
 *   > *"Then I need proper downloads folder and all of this stuff, history,
 *   > save passwords and all of this."*
 *
 * ## Why it is not in `settings.json`
 *
 * That store is the wrong shape and the wrong place, both. Wrong shape:
 * `settings-extra.ts` takes **primitives only**, capped at `MAX_KEYS = 500` and
 * `MAX_STRING_LENGTH = 4096`, and it says why — a settings file is a list of
 * small choices, not a document store. Wrong place: `settings.read` is a
 * copilot tool (`deck-control/catalogue.ts`), so an agent holding it can read
 * that whole bag. Where somebody has been is not a preference and it is not an
 * agent's business, so it lives in its own file beside `browser-profiles.json`,
 * which is where the thing it belongs to lives.
 *
 * ## Why a profile owns its history
 *
 * A profile here is a real Chromium partition — its own cookie jar, storage and
 * cache (`browser-profiles.ts`). Two profiles exist so two people, or two
 * personas, can use the same window without seeing each other's session, and a
 * shared history would hand back exactly what the partition keeps apart. So
 * every row carries the `profileId` it was recorded under and every read takes
 * one, the same discipline `browser-passwords.ts` applies to logins.
 *
 * **An Isolated tab records nothing.** Its partition is in memory and dies with
 * the tab (`browser-isolation.ts`), and its `profileId` is the empty string —
 * the same marker that stops a saved login ever being offered to it. A tab that
 * throws its cookies away and keeps a permanent record of where it went would be
 * a private mode that is not one.
 *
 * ## One row per address, not one per visit
 *
 * Chrome keeps every visit and groups them by day. This keeps the **last** visit
 * to each address in a profile, with a count. The store is a flat file read into
 * memory at first use, so the honest limit is the size of that file rather than
 * the size of a database, and one row per address makes {@link MAX_ENTRIES} a
 * number of *places* rather than a number of *reloads* — a page refreshed forty
 * times would otherwise push forty other sites out. Both facts are kept because
 * the address bar ranks on both: see {@link suggestFor}.
 */

/* ------------------------------------------------------------------ shape -- */

/** One address, in one profile, as it was last seen. */
export interface Visit {
  /** The profile whose partition the page was loaded in. Never `''`. */
  profileId: string
  url: string
  /** The page's own title, or `''` when it never announced one. */
  title: string
  /** Milliseconds since the epoch, of the most recent visit. */
  visitedAt: number
  /** How many times this address has been landed on. At least 1. */
  visits: number
}

/**
 * How many addresses a profile keeps.
 *
 * A number of places rather than a number of days: an expiry in days throws away
 * the site somebody visits once a year and keeps a thousand rows of a dev server
 * from this morning, and it is the first kind that is hard to find again. 3000
 * rows is around 300 kB of JSON, which is a file this module can read whole at
 * first use without anybody noticing.
 */
export const MAX_ENTRIES = 3000

/** Longer than this is a `data:` URL somebody pasted, not a page. */
export const MAX_URL = 2048

/** Longer than this is a document, not a title. */
export const MAX_TITLE = 300

/**
 * The address a visit is filed under, or null when it is not one to remember.
 *
 * http and https only — the same boundary `browser-passwords.ts` draws, and for
 * a related reason: `about:blank`, `data:` and the app's own start page are not
 * places, and a history whose first three rows are `about:blank` is a list
 * nobody reads twice. Plain http stays, because localhost is plain http and a
 * dev server is the most common page this browser will ever see.
 */
export function visitableUrl(url: unknown): string | null {
  if (typeof url !== 'string' || url.trim() === '' || url.length > MAX_URL) return null
  try {
    const parsed = new URL(url)
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return null
    if (parsed.hostname === '') return null
    return parsed.href
  } catch {
    return null
  }
}

/**
 * A title as it can be drawn on one row.
 *
 * Control characters are stripped rather than escaped, for the reason
 * `cleanProfileName` gives: a newline in the middle of a row is a rendering bug
 * whose cause is invisible to the person looking at it.
 */
export function cleanTitle(raw: unknown): string {
  if (typeof raw !== 'string') return ''
  const flat = raw.replace(/[\u0000-\u001f\u007f]/g, ' ').replace(/\s+/g, ' ').trim()
  return flat.length > MAX_TITLE ? flat.slice(0, MAX_TITLE) : flat
}

/**
 * Read a stored file into a list, dropping anything that is not a visit.
 *
 * Defensive even though this app wrote the file, the same discipline `readLogins`
 * and `readProfileState` apply: a row with no URL cannot be clicked, and a row
 * with no profile would appear in every profile's list — the one thing this
 * store exists to never do.
 */
export function readVisits(raw: unknown): Visit[] {
  if (typeof raw !== 'object' || raw === null) return []
  const value = raw as Record<string, unknown>
  const list = Array.isArray(value.entries) ? value.entries : []
  const out: Visit[] = []
  for (const entry of list) {
    if (typeof entry !== 'object' || entry === null) continue
    const record = entry as Record<string, unknown>
    const url = visitableUrl(record.url)
    const profileId = typeof record.profileId === 'string' ? record.profileId : ''
    if (url === null || profileId === '') continue
    out.push({
      profileId,
      url,
      title: cleanTitle(record.title),
      visitedAt: typeof record.visitedAt === 'number' ? record.visitedAt : 0,
      visits: typeof record.visits === 'number' && record.visits > 0 ? Math.floor(record.visits) : 1,
    })
  }
  return out
}

/**
 * Put a visit in, or move the row it already has.
 *
 * A title that arrives empty does **not** wipe the one already stored. Titles
 * arrive after the navigation that carries them — `did-navigate` fires with the
 * URL and `page-title-updated` some milliseconds later — so the first record of
 * any page has no title, and a redirect back onto a page would otherwise blank a
 * row that had a good one.
 *
 * Trimming happens here rather than at save time, so the list in memory and the
 * file on disk are never two different lengths.
 */
export function noteVisit(
  list: readonly Visit[],
  entry: { profileId: string; url: string; title: string; visitedAt: number },
  limit = MAX_ENTRIES,
): Visit[] {
  const existing = list.find((item) => item.profileId === entry.profileId && item.url === entry.url)
  const rest = existing ? list.filter((item) => item !== existing) : [...list]
  const next: Visit = {
    profileId: entry.profileId,
    url: entry.url,
    title: entry.title === '' && existing ? existing.title : entry.title,
    visitedAt: entry.visitedAt,
    visits: existing ? existing.visits + 1 : 1,
  }
  const all = [next, ...rest]
  if (all.length <= limit) return all
  // Oldest out of the door first, whatever profile it belongs to. A cap per
  // profile would let a profile nobody opens hold rows that a profile in daily
  // use has just lost.
  return [...all].sort((a, b) => b.visitedAt - a.visitedAt).slice(0, limit)
}

/** Everything matching `query` in one profile, newest first. */
export function historyFor(
  list: readonly Visit[],
  profileId: string,
  query = '',
  limit = 500,
): Visit[] {
  const needle = query.trim().toLowerCase()
  return list
    .filter((item) => item.profileId === profileId)
    .filter(
      (item) =>
        needle === '' ||
        item.url.toLowerCase().includes(needle) ||
        item.title.toLowerCase().includes(needle),
    )
    .sort((a, b) => b.visitedAt - a.visitedAt)
    .slice(0, limit)
}

/* ----------------------------------------------------------- the omnibox -- */

/**
 * The address as somebody would have typed it: whole, without its scheme, and
 * without `www.`.
 *
 * That is the whole judgement here — a person reaching for `github.com/asadev`
 * types `git`, never `htt`. These are the forms {@link scoreVisit} matches a
 * prefix against, and the forms the renderer completes inline.
 */
export function typedForms(url: string): string[] {
  const bare = url.replace(/^https?:\/\//i, '')
  const noWww = bare.replace(/^www\./i, '')
  return noWww === bare ? [url, bare] : [url, bare, noWww]
}

/**
 * How well a stored visit answers what is being typed, or 0 for not at all.
 *
 * Ordered by how confident the match is rather than by string distance, because
 * the top row is the one that gets completed inline and a wrong guess there
 * fights the person typing:
 *
 *  4. the address, shorn of scheme and `www.`, starts with what was typed —
 *     the row Chrome fills in for you.
 *  3. the whole address starts with it, so somebody who did type `https://`
 *     gets the same treatment.
 *  2. the title starts with it.
 *  1. either contains it anywhere — which is how a search run once before comes
 *     back from a word in the middle of its title.
 */
export function scoreVisit(item: Visit, typed: string): number {
  const needle = typed.trim().toLowerCase()
  if (needle === '') return 0
  const url = item.url.toLowerCase()
  const title = item.title.toLowerCase()
  const forms = typedForms(url)
  if (forms.slice(1).some((form) => form.startsWith(needle))) return 4
  if (url.startsWith(needle)) return 3
  if (title.startsWith(needle)) return 2
  if (url.includes(needle) || title.includes(needle)) return 1
  return 0
}

/**
 * What the address bar should offer for what has been typed so far.
 *
 * Score first, then how often the place has been visited, then how recently.
 * Visits before recency deliberately: an address bar is used to *return*
 * somewhere, and the page somebody opens every day beats the one they opened
 * once an hour ago even though the second is newer.
 */
export function suggestFor(
  list: readonly Visit[],
  profileId: string,
  typed: string,
  limit = 8,
): Visit[] {
  const scored: { item: Visit; score: number }[] = []
  for (const item of list) {
    if (item.profileId !== profileId) continue
    const score = scoreVisit(item, typed)
    if (score > 0) scored.push({ item, score })
  }
  scored.sort(
    (a, b) =>
      b.score - a.score || b.item.visits - a.item.visits || b.item.visitedAt - a.item.visitedAt,
  )
  return scored.slice(0, limit).map((entry) => entry.item)
}

/* ----------------------------------------------------------- persistence -- */

export function historyPath(userData: string): string {
  return join(userData, 'browser-history.json')
}

/**
 * In memory for the life of the process, written through on a short delay.
 *
 * Every navigation and every title change is a change, and a page whose title
 * ticks — a timer, an unread count — would otherwise be a file write a second,
 * forever. The delay coalesces those; {@link flushHistory} makes quitting exact.
 * A crash loses at most {@link WRITE_DELAY} of browsing, which is the right
 * trade against rewriting 300 kB because a title changed by one digit.
 */
const WRITE_DELAY = 400

let cache: Visit[] | null = null
let cacheDir: string | null = null
let pending: ReturnType<typeof setTimeout> | null = null

/** For tests, which must not inherit each other's store. */
export function resetHistoryForTests(): void {
  if (pending) clearTimeout(pending)
  pending = null
  cache = null
  cacheDir = null
}

export function allVisits(userData: string): Visit[] {
  if (cache !== null && cacheDir === userData) return cache
  cacheDir = userData
  const path = historyPath(userData)
  if (!existsSync(path)) {
    cache = []
    return cache
  }
  try {
    cache = readVisits(JSON.parse(readFileSync(path, 'utf8')) as unknown)
  } catch {
    // A file this app cannot read is the same as no file, from here. The
    // alternative is a browser panel that will not open because a JSON file has
    // a stray comma in it — the trap `readProfileState` names.
    cache = []
  }
  return cache
}

function write(userData: string, list: readonly Visit[]): void {
  const path = historyPath(userData)
  mkdirSync(dirname(path), { recursive: true })
  writeFileAtomic(path, JSON.stringify({ version: 1, entries: list }, null, 2))
}

function schedule(userData: string): void {
  if (pending) return
  pending = setTimeout(() => {
    pending = null
    try {
      write(userData, cache ?? [])
    } catch {
      // A history that could not be written is not worth a dialog: nothing else
      // in the app is blocked by it, and the next visit tries again.
    }
  }, WRITE_DELAY)
}

/** Write anything the delay is still holding. Called on the way out of the app. */
export function flushHistory(): void {
  if (pending) {
    clearTimeout(pending)
    pending = null
  }
  if (cache === null || cacheDir === null) return
  try {
    write(cacheDir, cache)
  } catch {
    /* see `schedule` */
  }
}

/**
 * Record a page that has just been landed on.
 *
 * Answers whether it was recorded, so its one caller does not have to carry a
 * second copy of this module's rules about which pages count.
 */
export function rememberVisit(
  userData: string,
  entry: { profileId: unknown; url: unknown; title?: unknown },
  now = Date.now(),
): boolean {
  // The empty profile id is an Isolated tab. See the note at the top of the file.
  if (typeof entry.profileId !== 'string' || entry.profileId === '') return false
  const url = visitableUrl(entry.url)
  if (url === null) return false
  cache = noteVisit(allVisits(userData), {
    profileId: entry.profileId,
    url,
    title: cleanTitle(entry.title),
    visitedAt: now,
  })
  cacheDir = userData
  schedule(userData)
  return true
}

/** Drop one address from one profile's history. */
export function forgetVisit(userData: string, profileId: unknown, url: unknown): Visit[] {
  const next = allVisits(userData).filter(
    (item) => !(item.profileId === profileId && item.url === url),
  )
  cache = next
  cacheDir = userData
  schedule(userData)
  return next
}

/** Drop a whole profile's history, and nobody else's. */
export function clearHistory(userData: string, profileId: unknown): Visit[] {
  const next = allVisits(userData).filter((item) => item.profileId !== profileId)
  cache = next
  cacheDir = userData
  schedule(userData)
  return next
}

/* -------------------------------------------------------------- register -- */

/**
 * Wire browsing history. Call once from `registerIpc()`:
 *
 *     import { registerBrowserHistoryIpc } from './browser-history'
 *     registerBrowserHistoryIpc(ipcMain, () => app.getPath('userData'))
 *
 * Channels:
 * - `browser-history:list`    (invoke, profileId, query?) → {@link Visit}[]
 * - `browser-history:suggest` (invoke, profileId, typed)  → {@link Visit}[]
 * - `browser-history:forget`  (invoke, profileId, url)    → that profile's list
 * - `browser-history:clear`   (invoke, profileId)         → that profile's list
 */
export function registerBrowserHistoryIpc(ipcMain: IpcMain, userData: () => string): void {
  const asId = (value: unknown): string => (typeof value === 'string' ? value : '')

  ipcMain.handle('browser-history:list', (_event, profileId: unknown, query: unknown) =>
    historyFor(allVisits(userData()), asId(profileId), typeof query === 'string' ? query : ''),
  )

  ipcMain.handle('browser-history:suggest', (_event, profileId: unknown, typed: unknown) =>
    suggestFor(allVisits(userData()), asId(profileId), typeof typed === 'string' ? typed : ''),
  )

  ipcMain.handle('browser-history:forget', (_event, profileId: unknown, url: unknown) =>
    historyFor(forgetVisit(userData(), asId(profileId), url), asId(profileId)),
  )

  ipcMain.handle('browser-history:clear', (_event, profileId: unknown) =>
    historyFor(clearHistory(userData(), asId(profileId)), asId(profileId)),
  )
}

File: ios/TerminalDeck/Ports/BrowserHistory.swift
/**
 * Every page this phone has opened through a tunnel, and when.
 *
 * Asad, listing what the browser on the phone was missing: *"search history and
 * cookies and all of this. Everything that Mac side had."* The cookies were
 * already there — `BrowserBridge` builds its web view on
 * `WKWebsiteDataStore.default()` precisely so a dev server that logs you in
 * keeps you logged in between taps. The history was not: `LocalhostBrowser` is
 * pushed, and the moment it is popped the tunnel closes, the web view is torn
 * down and everything it had been showing is gone. A browser that cannot tell
 * you where you were ten minutes ago is a window, not a browser.
 *
 * ## Why this is a file on the phone and not a frame on the wire
 *
 * Asad, drawing the line between the two halves of this app: *"any of the
 * feature which is not something that we can wire or it should be sitting in the
 * device whichever the device we are using at the moment and it is not related
 * to the folder sided or server sided then build also but keep it on the phone
 * side or app side."*
 *
 * A visit is a fact about **this phone**. The machine already knows what it
 * served — it has its own logs and its own browser history — and what it has no
 * business holding is a record of which of its pages somebody looked at from the
 * sofa. So nothing here reaches the desktop: no capability, no frame, no host
 * code. It is `UserDefaults`, the same as `PortBook` next door, and it behaves
 * the way Safari's own history behaves — it belongs to the device it was made
 * on.
 *
 * ## Keyed by machine, because a port number is not an address
 *
 * `localhost:3000` on his Mac and `localhost:3000` on a Hetzner box are two
 * unrelated pages that happen to share a number, and a history keyed on the
 * number alone would offer one machine's rows while the other is connected — and
 * then open something else entirely when one was tapped. `PortBook` keys on
 * `host:` for exactly this reason and this does the same, against the same
 * stable `DeckEndpoint.hostId`.
 *
 * ## What a visit is, and why it is a port and a path rather than a URL string
 *
 * The URL the web view reports is `http://127.0.0.1:3000/admin` — or
 * `http://[::1]:3000/admin`, because `PortTunnel` binds whichever loopback
 * family won the race, and in the Simulator that is routinely the v6 one. Two
 * spellings of one page. Storing the raw string would put the same page on the
 * list twice, once per literal, and neither row would be wrong.
 *
 * So a visit holds the **port on the machine** and the **path**, which is the
 * pair that identifies the page and — not by coincidence — the pair
 * `MachineBrowserView.openHere(_:_:)` needs to open it again. The address a row
 * draws is derived from them: `localhost:3000/admin`, which is what somebody
 * would have typed into the address bar to get there. The tunnel binds the
 * machine's own port number on this phone's loopback, so the number in the URL
 * *is* the number on the machine — see `PortTunnel.bind`.
 *
 * ## The title arrives second, so recording is two calls and not one
 *
 * A document has no title until it has loaded. `WKWebView` publishes the new
 * `url` at the start of a navigation and the new `title` some time after it —
 * and in between, the title property still holds the **previous page's** title.
 * A single `record(url, title)` called from both signals would therefore write
 * the last page's name onto this page's row, and the row somebody scrolls back
 * to a week later would be a lie with a plausible name on it.
 *
 * Hence two entry points with two different claims. `record(address:host:)` says
 * *this page was opened* and makes no claim about its name; `retitle` says *the
 * document that is on screen calls itself this*. An empty title never overwrites
 * a real one, which is what makes the clearing-then-setting `WKWebView` does on
 * every navigation harmless.
 *
 * ## The title is the page's text, not the user's
 *
 * This is the one place this store differs from `PortBook` in kind rather than
 * in shape. A port's name is typed by the person holding the phone; a page title
 * is `document.title` on a site that came down a tunnel, and a page can set that
 * to ten thousand characters of anything it likes. So it is bounded and stripped
 * of control characters on the way in *and* on the way back off disk, for the
 * same reason `DevServerReport.note` is treated as untrusted: it lands on a row
 * in this app's own chrome.
 */

import Foundation
import Observation

/**
 * Deliberately **not** `@MainActor`, for the reason written out at length in
 * `PortBook`: a screen holds one of these as `var history: BrowserHistory =
 * .shared`, that is a default argument on a memberwise initialiser, and default
 * arguments are evaluated in a non-isolated context — so a main-actor `shared`
 * could not be named there at all. Nothing in here touches UIKit or a socket;
 * it is a dictionary and a `UserDefaults` write, and every caller is a view or a
 * bridge already on the main thread.
 */
@Observable
final class BrowserHistory {

    /**
     * The one the screens read.
     *
     * A singleton beside `PortBook.shared` rather than something hung off
     * `DeckModel`, and for the same reason: this is a property of *this phone*
     * rather than of a machine. A phone that is paired with nothing still has a
     * history of the machines it used to be paired with, and a model rebuilt on
     * a reconnect must not take it with it.
     */
    static let shared = BrowserHistory()

    /**
     * The longest history one machine keeps.
     *
     * Two hundred **distinct pages** — the list is deduplicated, so this is not
     * two hundred navigations, it is two hundred different addresses, which on a
     * dev machine is weeks of work rather than an afternoon.
     *
     * The number is bounded by where it lives rather than by what anybody wants
     * to scroll. `UserDefaults` is read into memory whole at launch and this
     * store is written whole on every visit; at roughly 120 bytes of JSON per
     * row that is 24KB per machine, which is nothing to load and nothing to
     * write. Ten thousand rows would be neither, and the screen that reads them
     * is searched rather than scrolled to the end — nobody is going to find the
     * four-thousandth row by dragging.
     */
    static let maxVisits = 200

    /**
     * The longest page title a row will keep.
     *
     * A row draws one line and a phone runs out somewhere past forty characters,
     * so this is not about drawing — it is about what a **search** should still
     * be able to find. A title cut at forty would drop the end of *"Orders ·
     * Admin · Storefront (staging)"*, which is the half somebody searches for.
     * A hundred and twenty keeps every real title and refuses the pathological
     * one a page could hand over.
     */
    static let maxTitleLength = 120

    /**
     * The longest path this will remember, and it **refuses** rather than cuts.
     *
     * Truncating a path produces an address that is not the page — a row that
     * would open something else when it was tapped, which is worse than not
     * having the row. So an absurd URL is simply not recorded. A kilobyte is far
     * past any real dev-server route, including the query strings a dashboard
     * puts its filters in.
     */
    static let maxPathLength = 1024

    /**
     * How long a repeat visit to the page already at the top is folded into the
     * one before it without touching the disk.
     *
     * The case this exists for is the one this whole feature is aimed at: a
     * dev server with hot reload navigates the same URL dozens of times a minute
     * while somebody saves a file. Deduplication already means those are one
     * row — but each of them would still re-encode the whole store and write it,
     * for a timestamp moving by a second and a half. Two seconds of drift
     * between the time in memory and the time on disk costs a row that says
     * *2 minutes ago* when it could have said *2 minutes ago*.
     */
    static let coalesce: TimeInterval = 2

    /**
     * One page, once.
     *
     * `port` and `path` are the identity and the answer both: they are what
     * makes two spellings of a loopback address one row, and they are exactly
     * what `MachineBrowserView.openHere(_:_:)` takes.
     */
    struct Visit: Identifiable, Equatable, Codable {
        /// The port **on the machine**. The tunnel binds the same number on this
        /// phone, so it is also the number in the URL the web view reported.
        let port: Int
        /// Always begins with `/`, and carries the query and the fragment when
        /// there were any — the same string `LocalhostAddress` produces and the
        /// same one `LocalhostBrowser` resolves against the tunnel's origin.
        let path: String
        /// What the document called itself, or empty until it said. Empty is a
        /// normal state, not a missing one — see the type header.
        var title: String
        var at: Date

        /// Unique within one machine's list, because that is what deduplication
        /// means here.
        var id: String { "\(port)\(path)" }

        /**
         * What the row draws, and what somebody would have typed to get here.
         *
         * The trailing slash goes for a root path so that this reads the same as
         * the port rows one screen over — `PortRow` draws `localhost:3000` — and
         * so the two screens do not appear to be talking about different things.
         * `LocalhostAddress.parse` accepts both spellings, so nothing downstream
         * cares which one is shown.
         */
        var address: String {
            path == "/" ? "localhost:\(port)" : "localhost:\(port)\(path)"
        }
    }

    private let defaults: UserDefaults
    /// The clock, as a seam. The coalescing window above is the only piece of
    /// behaviour here with a duration in it, and a test proving it must be able
    /// to move time rather than sleep through it.
    private let now: () -> Date
    private static let storageKey = "terminaldeck.browserHistory.v1"

    /// host id → visits, newest first. Held in order rather than sorted on read
    /// because every read is a screen drawing and every write is one insertion.
    private var byHost: [String: [Visit]] = [:]

    /// `defaults` is a seam for the tests, which run against their own suite so
    /// a test run cannot write history onto the machine it is running from —
    /// the same arrangement `PortBook` uses.
    init(defaults: UserDefaults = .standard, now: @escaping () -> Date = Date.init) {
        self.defaults = defaults
        self.now = now
        load()
    }

    // MARK: - Reading

    /// One machine's history, newest first.
    func visits(host: String) -> [Visit] {
        byHost[host] ?? []
    }

    /**
     * One machine's history, filtered by what somebody typed.
     *
     * **Every whitespace-separated word has to appear somewhere**, in the
     * address or in the title, rather than the whole query having to appear as
     * one substring. Somebody looking for the settings page on port 3000 types
     * `3000 settings`, and no single string on that row contains that — the
     * number is in the address and the word is in the title. A plain
     * `contains(query)` would find nothing and the screen would look empty and
     * broken while the row it was being asked for sat one line below the field.
     *
     * `localizedStandardContains` rather than `lowercased().contains`, so the
     * search is case- **and** diacritic-insensitive and behaves the way search
     * behaves everywhere else on the phone.
     */
    func visits(host: String, matching query: String) -> [Visit] {
        let terms = query.split(whereSeparator: \.isWhitespace).map(String.init)
        guard !terms.isEmpty else { return visits(host: host) }
        return visits(host: host).filter { visit in
            terms.allSatisfy { term in
                visit.address.localizedStandardContains(term)
                    || visit.title.localizedStandardContains(term)
            }
        }
    }

    // MARK: - Writing

    /**
     * A page was opened. Nothing is claimed about its name.
     *
     * `address` is whatever the web view currently reports — a full
     * `http://127.0.0.1:3000/admin`. It is put through `LocalhostAddress`, which
     * is the one place in this app that decides what a loopback address is, and
     * anything it refuses is dropped without a sound: `about:blank`, a `data:`
     * URL and the error page a failed load leaves behind are all things that
     * happen on this screen and none of them is a page somebody visited.
     *
     * A page already in the list keeps **its own** title and moves to the top
     * with a new time. That is not the stale-title problem the header describes
     * — it is the opposite, and it is what makes a page you have been to before
     * come back with its name already on it instead of blank until it reloads.
     */
    func record(address: String, host: String) {
        guard !host.isEmpty else { return }
        guard case let .address(port, path) = LocalhostAddress.parse(address) else { return }
        guard path.count <= Self.maxPathLength else { return }

        var list = byHost[host] ?? []
        let stamp = now()
        let id = "\(port)\(path)"

        if let first = list.first, first.id == id {
            // The hot-reload case. Same page, still on top, seen a moment ago:
            // move the clock in memory and leave the disk alone.
            guard stamp.timeIntervalSince(first.at) >= Self.coalesce else {
                list[0].at = stamp
                byHost[host] = list
                return
            }
            list[0].at = stamp
            byHost[host] = list
            save()
            return
        }

        var title = ""
        if let existing = list.firstIndex(where: { $0.id == id }) {
            title = list[existing].title
            list.remove(at: existing)
        }
        list.insert(Visit(port: port, path: path, title: title, at: stamp), at: 0)
        byHost[host] = Self.bounded(list)
        save()
    }

    /**
     * The document on screen says what it is called.
     *
     * Addressed by the page rather than by an index, because the caller is a
     * KVO notification about a web view and has no idea which row it belongs to
     * — it knows the address the view is showing, and that is enough.
     *
     * An empty or whitespace-only title is **ignored**, never written. WebKit
     * blanks `title` on every navigation before the new document sets its own,
     * and a store that took that literally would wipe the name off a row a
     * fraction of a second after writing it.
     *
     * A title for a page that is not in the list yet records the visit as well.
     * It should not be reachable — the address changes before the title does —
     * but the alternative is silently dropping the only name a page will ever
     * give, and the visit it implies is real either way.
     */
    func retitle(address: String, title: String, host: String) {
        guard !host.isEmpty, let cleaned = Self.clean(title) else { return }
        guard case let .address(port, path) = LocalhostAddress.parse(address) else { return }
        guard path.count <= Self.maxPathLength else { return }

        var list = byHost[host] ?? []
        let id = "\(port)\(path)"
        if let index = list.firstIndex(where: { $0.id == id }) {
            guard list[index].title != cleaned else { return }
            list[index].title = cleaned
        } else {
            list.insert(Visit(port: port, path: path, title: cleaned, at: now()), at: 0)
        }
        byHost[host] = Self.bounded(list)
        save()
    }

    // MARK: - Forgetting

    /// Drop one page from one machine's history. The swipe on the row.
    func forget(_ visit: Visit, host: String) {
        guard var list = byHost[host] else { return }
        list.removeAll { $0.id == visit.id }
        if list.isEmpty { byHost.removeValue(forKey: host) } else { byHost[host] = list }
        save()
    }

    /**
     * Drop everything this phone remembers about one machine.
     *
     * **One machine, not all of them**, and the screen names which one on the
     * button that does it. The screen is a single machine's history — its rows,
     * its search, its swipes — and a Clear on it that silently also wiped the
     * history of a laptop that is not even connected would be the app doing
     * something nobody could see it do. Somebody who wants both clears both,
     * from the two screens that show them.
     */
    func clear(host: String) {
        guard byHost[host] != nil else { return }
        byHost.removeValue(forKey: host)
        save()
    }

    // MARK: - Bounds

    /**
     * Newest first, no duplicates, and no longer than the cap — oldest dropped.
     *
     * Applied on the way in and again on the way back off disk. The sort is not
     * redundant with insertion order: `retitle` can create a row for a page
     * whose time is older than the one above it, and a file written by an older
     * build has whatever order it has.
     *
     * **Ties keep the order they came in**, which is the position in the array,
     * which is newest first. Swift's sort does not promise to be stable, and two
     * visits carrying the identical instant is not a theoretical case: it is
     * every test that injects a fixed clock, and a screen whose rows shuffled
     * between two runs would make one of those tests fail once a fortnight for
     * no reason anybody could reproduce.
     */
    private static func bounded(_ visits: [Visit]) -> [Visit] {
        var seen = Set<String>()
        let unique = visits
            .enumerated()
            .sorted { left, right in
                left.element.at == right.element.at
                    ? left.offset < right.offset
                    : left.element.at > right.element.at
            }
            .map(\.element)
            .filter { seen.insert($0.id).inserted }
        guard unique.count > maxVisits else { return unique }
        return Array(unique.prefix(maxVisits))
    }

    /**
     * A page title as it will be stored, or nil when there is nothing left.
     *
     * The same shape as `PortBook.clean` and deliberately **not** the same
     * function: that one enforces a forty-character bound on a name somebody
     * typed, this one enforces a much longer bound on text a web page handed
     * over, and sharing the code would mean a change to what fits on a port row
     * silently truncating everybody's browsing history.
     *
     * Control characters go rather than being folded to spaces — a title with a
     * newline in it is a page being odd, and joining the halves would invent a
     * string the page never had.
     */
    static func clean(_ raw: String?) -> String? {
        guard let raw else { return nil }
        let stripped = raw.unicodeScalars
            .filter { !CharacterSet.controlCharacters.contains($0) }
            .reduce(into: "") { $0.unicodeScalars.append($1) }
        let trimmed = stripped.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        guard trimmed.count > maxTitleLength else { return trimmed }
        return String(trimmed.prefix(maxTitleLength)).trimmingCharacters(in: .whitespaces)
    }

    // MARK: - Storage

    private struct Stored: Codable {
        var visits: [String: [Visit]]
    }

    /**
     * Read back, and treated as untrusted on the way out.
     *
     * Every bound is re-applied. Half of what is in this file arrived from a web
     * page, the file itself is editable by hand in a Simulator, and a record
     * written by a build with a different cap must not be able to get around
     * this one's. A row that does not survive the checks is dropped rather than
     * repaired: a port outside the legal range or a path that does not start at
     * the root is not a page this app could open, so a row offering to open it
     * would be a button that cannot work.
     */
    private func load() {
        guard let data = defaults.data(forKey: Self.storageKey),
              let stored = try? JSONDecoder().decode(Stored.self, from: data) else { return }
        byHost = stored.visits.compactMapValues { list -> [Visit]? in
            let usable = list.compactMap { visit -> Visit? in
                guard visit.port > 0, visit.port <= 65_535 else { return nil }
                guard visit.path.hasPrefix("/"), visit.path.count <= Self.maxPathLength else {
                    return nil
                }
                var row = visit
                row.title = Self.clean(visit.title) ?? ""
                return row
            }
            return usable.isEmpty ? nil : Self.bounded(usable)
        }
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(Stored(visits: byHost)) else { return }
        defaults.set(data, forKey: Self.storageKey)
    }
}
